src/Background/NotificationRequest.vala

```diff
@@ -0,0 +1,120 @@
+        CANCELLED,
+        FAILED
+    }
+
+    [DBus (name = "org.freedesktop.Notifications")]
+    public interface Notifications : Object {
+        public signal void action_invoked (uint32 id, string action_key);
+        public signal void notification_closed (uint32 id, uint32 reason);
+
+        public abstract void close_notification (uint32 id) throws DBusError, IOError;
+        public abstract uint32 notify (
+            string app_name,
+            uint32 replaces_id,
+            string app_icon,
+            string summary,
+            string body,
+            string[] actions,
+            HashTable<string, Variant> hints,
+            int32 expire_timeout
+        ) throws DBusError, IOError;
+    }
+
+    [DBus (visible = false)]
+    public static void init (DBusConnection connection) {
+        notification_by_id = new HashTable<uint32, NotificationRequest> (null, null);
+
+        try {
+            notifications = connection.get_proxy_sync ("org.freedesktop.Notifications", "/org/freedesktop/Notifications");
+            notifications.action_invoked.connect (on_action_invoked);
+            notifications.notification_closed.connect (on_notification_closed);
+        } catch {
+            warning ("Cannot connect to notifications dbus, background portal working with reduced functionality.");
+        }
+    }
+
+    private static void on_action_invoked (uint32 id, string action_key) {
+        var notification = notification_by_id.take (id);
+        if (notification == null) {
+            return;
+        }
+
+        if (action_key == ACTION_ALLOW_BACKGROUND) {
+            notification.response (NotifyBackgroundResult.ALLOW);
+        } else {
+            notification.response (NotifyBackgroundResult.FORBID);
+        }
+    }
+
+    private static void on_notification_closed (uint32 id, uint32 reason) {
+        var notification = notification_by_id.take (id);
+        if (notification == null) {
+            return;
+        }
+
+        if (reason == 2) { //Dismissed by user
+            notification.response (NotifyBackgroundResult.ALLOW_ONCE);
+        } else if (reason == 3) { //Closed via DBus call
+            notification.response (NotifyBackgroundResult.CANCELLED);
+        }
+    }
+
+    [DBus (visible = false)]
+    public void send_notification (string app_id, string app_name) {
+        string[] actions = {
+            ACTION_ALLOW_BACKGROUND,
+            _("Allow"),
+            ACTION_FORBID_BACKGROUND,
+            _("Forbid")
+        };
+        var hints = new HashTable<string, Variant> (null, null);
+        hints["desktop-entry"] = app_id;
+        hints["urgency"] = (uint8) 1;
+
+        try {
+            id = notifications.notify (
+                app_name, 0, "",
+                _("Background activity"),
+                _(""""%s" is running in the background without appropriate permission""").printf (app_name),
+                actions,
+                hints,
+                0
```

Comment:
the spec says, a 0 timeout is infinite and -1 is to let the timeout to the server, the elementary one doesn't respect the sent timeout right now, but i plan to fix it.

would be good to let the timeout to the server. instead of sending an non-expiring notification.

```suggestion
                -1
```